File: app/src/main/java/movies/test/softserve/movies/activity/TVShowDetailsActivity.java
package movies.test.softserve.movies.activity;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.arch.lifecycle.ViewModelProviders;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.Point;
import android.graphics.Rect;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.Gravity;
import android.view.View;
import android.view.animation.DecelerateInterpolator;
import android.widget.ImageView;
import android.widget.RatingBar;
import android.widget.TextView;

import com.facebook.share.model.ShareLinkContent;
import com.facebook.share.widget.ShareDialog;
import com.squareup.picasso.Picasso;
import com.squareup.picasso.Target;

import movies.test.softserve.movies.R;
import movies.test.softserve.movies.adapter.HorizontalImageAdapter;
import movies.test.softserve.movies.entity.FullTVShow;
import movies.test.softserve.movies.entity.TVEntity;
import movies.test.softserve.movies.event.OnFullTVShowGetListener;
import movies.test.softserve.movies.event.OnInfoUpdatedListener;
import movies.test.softserve.movies.repository.TVShowsRepository;
import movies.test.softserve.movies.service.DBHelperService;
import movies.test.softserve.movies.service.DBMovieService;
import movies.test.softserve.movies.viewmodel.FullTVSeriesViewModel;

/**
 * Created by rkrit on 20.11.17.
 */

public class TVShowDetailsActivity extends BaseActivity {
    public static final String TV_ENTITY = "tv entity";

    private TVEntity tvShow;
    private FullTVSeriesViewModel viewModel;

    private DBHelperService helperService = new DBHelperService();
    private DBMovieService dbService = DBMovieService.getInstance();
    private TVShowsRepository repository = TVShowsRepository.getInstance();

    private Animator mCurrentAnimator;
    private BitmapDrawable mBitmapDrawable;
    private int mShortAnimationDuration;
    private OnFullTVShowGetListener listenerW;
    private OnInfoUpdatedListener onInfoUpdatedListener;

    private Toolbar toolbar;
    private FloatingActionButton fab;
    private ImageView watched;
    private CollapsingToolbarLayout toolbarLayout;
    private RatingBar ratingBar;
    private TextView voteCountView;
    private TextView overviewView;
    private TextView releaseDateView;
    private ImageView share;
    private TextView links;
    private RecyclerView seasons;


    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        viewModel = ViewModelProviders.of(this).get(FullTVSeriesViewModel.class);

        mShortAnimationDuration = getResources().getInteger(android.R.integer.config_mediumAnimTime);
        initView();
        getIntentInfo();
        useIntentInfo();
        if (savedInstanceState != null) {
            if (viewModel.getFullTVShow() != null)
                getFullInfo(viewModel.getFullTVShow());
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (listenerW == null) {
            if (viewModel.getFullTVShow() == null) {
                listenerW = new OnFullTVShowGetListener() {
                    @Override
                    public void onFullTVShowGet(FullTVShow tvShow) {
                        viewModel.setFullTVShow(tvShow);
                        getFullInfo(tvShow);
                    }
                };
                repository.addOnFullTVShowGetListeners(listenerW);
                repository.trytoGetFullTVShow(tvShow.getId());
            }
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (listenerW != null) {
            repository.removeOnFullTVShowGetListeners(listenerW);
            listenerW = null;
        }
        if (onInfoUpdatedListener != null) {
            repository.removeOnInfoUpdatedListener(onInfoUpdatedListener);
            onInfoUpdatedListener = null;
        }
    }

    private void getFullInfo(final FullTVShow fullTVShow) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                releaseDateView.setVisibility(View.VISIBLE);
                releaseDateView.setText(getString(R.string.release_date, fullTVShow.getFirstAirDate()));
                if (fullTVShow.getHomepage() != null && !fullTVShow.getHomepage().equals("")) {
                    links.setText(fullTVShow.getHomepage());
                    links.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            Uri webPage = Uri.parse(fullTVShow.getHomepage());
                            Intent webIntent = new Intent(Intent.ACTION_VIEW, webPage);
                            startActivity(webIntent);
                        }
                    });
                }
                seasons.setLayoutManager(new LinearLayoutManager(TVShowDetailsActivity.this,
                        LinearLayoutManager.HORIZONTAL, false));
                seasons.setAdapter(new HorizontalImageAdapter(fullTVShow.getSeasons(),
                        (image) -> zoomImageFromThumb(image,
                                ((BitmapDrawable) image.getDrawable()).getBitmap())));
            }
        });

    }


    private void useIntentInfo() {
        toolbarLayout = findViewById(R.id.toolbar_layout);
        toolbarLayout.setTitle(tvShow.getTitle());
        ratingBar = findViewById(R.id.ratingBar);
        toolbarLayout.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (mBitmapDrawable != null) {
                    zoomImageFromThumb(v, mBitmapDrawable.getBitmap());
                }
            }
        });
        ratingBar.setRating(tvShow.getVoteAverage().floatValue() / 2);
        ratingBar.setOnRatingBarChangeListener(new RatingBar.OnRatingBarChangeListener() {
            @Override
            public void onRatingChanged(final RatingBar ratingBar, float rating, boolean fromUser) {
                if (fromUser) {
                    onInfoUpdatedListener = new OnInfoUpdatedListener() {
                        @Override
                        public void OnInfoUpdated(float rating) {
                            ratingBar.setRating(rating);
                            Snackbar.make(findViewById(R.id.nested_scroll_view), getString(R.string.rating_saved), Snackbar.LENGTH_LONG).show();
                        }
                    };
                    repository.addOnInfoUpdatedListener(onInfoUpdatedListener);
                    repository.rateTVShow(tvShow.getId(), rating * 2);
                }
            }
        });


        voteCountView.setText(getString(R.string.vote_count, tvShow.getVoteAverage(), tvShow.getVoteCount()));
        overviewView.setText(tvShow.getOverview());
        releaseDateView.setVisibility(View.GONE);
        share.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                ShareLinkContent shareLinkContent = new ShareLinkContent.Builder()
                        .setQuote(tvShow.getTitle() + "     \r\nPlot: " + tvShow.
                                getOverview())
                        .setContentUrl(Uri.parse("https://image.tmdb.org/t/p/w500" + tvShow.
                                getPosterPath()))
                        .build();
                ShareDialog.show(TVShowDetailsActivity.this, shareLinkContent);
            }
        });
        Picasso
                .with(this)
                .load("https://image.tmdb.org/t/p/w500" + tvShow.getPosterPath())
                .into(new Target() {
                    @Override
                    public void onBitmapLoaded(Bitmap bitmap, Picasso.LoadedFrom from) {
                        mBitmapDrawable = new BitmapDrawable(getResources(), bitmap);
                        mBitmapDrawable.setGravity(Gravity.NO_GRAVITY);
                        toolbarLayout.setBackground(mBitmapDrawable);
                    }

                    @Override
                    public void onBitmapFailed(Drawable errorDrawable) {
                        Log.d("TAG", "FAILED");
                    }

                    @Override
                    public void onPrepareLoad(Drawable placeHolderDrawable) {
                        Log.d("TAG", "Prepare Load");
                    }
                });
    }

    private void getIntentInfo() {
        if (getIntent() != null && tvShow == null) {
            tvShow = (TVEntity) getIntent().getExtras().getSerializable(TV_ENTITY);
        }
        watched.setImageResource((dbService.checkIfExists(tvShow.getId())) ? R.mipmap.checked : R.mipmap.not_checked);
        watched.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                switch (helperService.toDoWithWatched(tvShow)) {
                    case WATCHED:
                        watched.setImageResource(R.mipmap.checked);
                        Snackbar.make(findViewById(R.id.nested_scroll_view), getString(R.string.added_to_watched), Snackbar.LENGTH_SHORT).show();
                        break;
                    case FAVOURITE:
                        Snackbar.make(findViewById(R.id.nested_scroll_view), getString(R.string.you_cant_favourrite), Snackbar.LENGTH_SHORT).show();
                        break;
                    case CANCELED:
                        AlertDialog.Builder builder = new AlertDialog.Builder(TVShowDetailsActivity.this);
                        builder.setMessage(getString(R.string.confirm))
                                .setPositiveButton(getString(R.string.yes), new DialogInterface.OnClickListener() {
                                    @Override
                                    public void onClick(DialogInterface dialog, int which) {
                                        dbService.deleteFromDb(tvShow.getId());
                                        watched.setImageResource(R.mipmap.not_checked);
                                        Snackbar.make(findViewById(R.id.nested_scroll_view), getString(R.string.mark_unwatched), Snackbar.LENGTH_SHORT).show();
                                    }
                                }).setNegativeButton(R.string.no, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                            }
                        });
                        builder.create().show();
                        break;

                }
            }
        });
        if (dbService.checkIfIsFavourite(tvShow.getId())) {
            fab.setImageResource(R.drawable.ic_stars_black_24dp);
        }
    }

    private void initView() {
        setContentView(R.layout.activity_movie_details);
        seasons = findViewById(R.id.genres);
        links = findViewById(R.id.links);
        toolbar = findViewById(R.id.toolbar);
        fab = findViewById(R.id.fab);
        watched = findViewById(R.id.watched);
        voteCountView = findViewById(R.id.vote_count);
        overviewView = findViewById(R.id.overview);
        releaseDateView = findViewById(R.id.release_date);
        share = findViewById(R.id.share);
        setSupportActionBar(toolbar);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (helperService.toDoWithFavourite(tvShow)) {
                    fab.setImageResource(R.drawable.ic_stars_black_24dp);
                    Snackbar.make(v, getString(R.string.added_to_favourite), Snackbar.LENGTH_SHORT)
                            .show();
                    watched.setImageResource(R.mipmap.checked);
                } else {
                    fab.setImageResource(R.drawable.ic_star_border_black_24dp);
                    Snackbar.make(v, getString(R.string.removed_from_favourite), Snackbar.LENGTH_SHORT)
                            .show();
                }
            }
        });
    }

    private void zoomImageFromThumb(final View thumbView, Bitmap bitmap) {
        if (mCurrentAnimator != null) {
            mCurrentAnimator.cancel();
        }

        final ImageView expandedImageView = findViewById(R.id.expanded_image);
        expandedImageView.setImageBitmap(bitmap);

        final Rect startBounds = new Rect();
        final Rect finalBounds = new Rect();
        final Point globalOffset = new Point();

        thumbView.getGlobalVisibleRect(startBounds);
        findViewById(R.id.container).getGlobalVisibleRect(finalBounds, globalOffset);
        startBounds.offset(-globalOffset.x, -globalOffset.y);
        finalBounds.offset(-globalOffset.x, -globalOffset.y);

        float startScale;
        if ((float) finalBounds.width() / finalBounds.height()
                > (float) startBounds.width() / startBounds.height()) {
            startScale = (float) startBounds.height() / finalBounds.height();
            float startWidth = startScale * finalBounds.width();
            float deltaWidth = (startWidth - startBounds.width()) / 2;
            startBounds.left -= deltaWidth;
            startBounds.right += deltaWidth;
        } else {
            startScale = (float) startBounds.width() / finalBounds.width();
            float startHeight = startScale * finalBounds.height();
            float deltaHeight = (startHeight - startBounds.height()) / 2;
            startBounds.top -= deltaHeight;
            startBounds.bottom += deltaHeight;
        }

        thumbView.setAlpha(0f);
        expandedImageView.setVisibility(View.VISIBLE);

        expandedImageView.setPivotX(0f);
        expandedImageView.setPivotY(0f);

        AnimatorSet set = new AnimatorSet();
        set
                .play(ObjectAnimator.ofFloat(expandedImageView, View.X, startBounds.left,
                        finalBounds.left))
                .with(ObjectAnimator.ofFloat(expandedImageView, View.Y, startBounds.top,
                        finalBounds.top))
                .with(ObjectAnimator.ofFloat(expandedImageView, View.SCALE_X, startScale, 1f))
                .with(ObjectAnimator.ofFloat(expandedImageView, View.SCALE_Y, startScale, 1f));
        set.setDuration(mShortAnimationDuration);
        set.setInterpolator(new DecelerateInterpolator());
        set.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animation) {
                mCurrentAnimator = null;
            }

            @Override
            public void onAnimationCancel(Animator animation) {
                mCurrentAnimator = null;
            }
        });
        set.start();
        mCurrentAnimator = set;

        final float startScaleFinal = startScale;
        expandedImageView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (mCurrentAnimator != null) {
                    mCurrentAnimator.cancel();
                }

                AnimatorSet set = new AnimatorSet();
                set
                        .play(ObjectAnimator.ofFloat(expandedImageView, View.X, startBounds.left))
                        .with(ObjectAnimator.ofFloat(expandedImageView, View.Y, startBounds.top))
                        .with(ObjectAnimator
                                .ofFloat(expandedImageView, View.SCALE_X, startScaleFinal))
                        .with(ObjectAnimator
                                .ofFloat(expandedImageView, View.SCALE_Y, startScaleFinal));
                set.setDuration(mShortAnimationDuration);
                set.setInterpolator(new DecelerateInterpolator());
                set.addListener(new AnimatorListenerAdapter() {
                    @Override
                    public void onAnimationEnd(Animator animation) {
                        thumbView.setAlpha(1f);
                        expandedImageView.setVisibility(View.GONE);
                        mCurrentAnimator = null;
                    }

                    @Override
                    public void onAnimationCancel(Animator animation) {
                        thumbView.setAlpha(1f);
                        expandedImageView.setVisibility(View.GONE);
                        mCurrentAnimator = null;
                    }
                });
                set.start();
                mCurrentAnimator = set;
            }
        });
    }
}

File: app/src/main/java/movies/test/softserve/movies/activity/TVShowDetailsActivity.kt
package movies.test.softserve.movies.activity

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.arch.lifecycle.ViewModelProviders
import android.content.DialogInterface
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.Point
import android.graphics.Rect
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.net.Uri
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.Gravity
import android.view.View
import android.view.animation.DecelerateInterpolator
import android.widget.ImageView
import com.facebook.share.model.ShareLinkContent
import com.facebook.share.widget.ShareDialog
import com.squareup.picasso.Picasso
import com.squareup.picasso.Target
import kotlinx.android.synthetic.main.activity_movie_details.*
import kotlinx.android.synthetic.main.content_movie_details.*
import movies.test.softserve.movies.R
import movies.test.softserve.movies.entity.FullTVShow
import movies.test.softserve.movies.entity.TVShow
import movies.test.softserve.movies.event.OnFullTVShowGetListener
import movies.test.softserve.movies.event.OnInfoUpdatedListener
import movies.test.softserve.movies.repository.TVShowsRepository
import movies.test.softserve.movies.service.DBMovieService
import movies.test.softserve.movies.viewmodel.FullTVSeriesViewModel

class TVShowDetailsActivity : AppCompatActivity() {

    companion object {
        val ID = "id"
        val NAME = "name"
        val VOTE_AVERAGE = "vote average"
        val VOTE_COUNT = "vote count"
        val POSTER_PATH = "poster path"
        val OVERVIEW = "overview"
    }

    private var viewModel: FullTVSeriesViewModel = FullTVSeriesViewModel()

    private var mCurrentAnimator: Animator? = null
    private var mBitmapDrawable: BitmapDrawable? = null

    private var mShortAnimationDuration: Int = 0

    private var listener : OnFullTVShowGetListener? = null
    private var onInfoUpdatedListener : OnInfoUpdatedListener? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewModel = ViewModelProviders.of(this).get(FullTVSeriesViewModel::class.java)

        mShortAnimationDuration = resources.getInteger(android.R.integer.config_shortAnimTime)
        initView()
        getIntentInfo()
        useIntentInfo()
        if (savedInstanceState!=null){
            if (viewModel.fullTVShow!=null)
                getFullInfo(viewModel.fullTVShow!!)
        }
    }

    override fun onResume() {
        super.onResume()
        if (listener==null){
            if (viewModel.fullTVShow==null) {
                listener = object : OnFullTVShowGetListener {
                    override fun onFullTVShowGet(tvShow: FullTVShow) {
                        viewModel.fullTVShow = tvShow
                        getFullInfo(tvShow)
                    }
                }
                TVShowsRepository.getInstance().addOnFullTVShowGetListeners(listener)
                TVShowsRepository.getInstance().trytoGetFullTVShow(viewModel.tvShow!!.id)
            }
        }
    }

    override fun onPause() {
        super.onPause()
        if (listener!=null){
            TVShowsRepository.getInstance().removeOnFullTVShowGetListeners(listener)
            listener = null
        }
        if (onInfoUpdatedListener!=null){
            TVShowsRepository.getInstance().removeOnInfoUpdatedListener(onInfoUpdatedListener!!)
            listener = null
        }
    }
    private fun initView() {
        setContentView(R.layout.activity_movie_details)
        setSupportActionBar(toolbar)
        fab.setOnClickListener { view ->
            val dbService = DBMovieService.getInstance()
            if (!dbService.checkIfIsFavourite(viewModel.tvShow!!.id)) {
                if (!dbService.checkIfExists(viewModel.tvShow!!.id)) {
                    dbService.insertTVShowToFavourite(viewModel.tvShow!!.id,
                            viewModel.tvShow!!.title,
                            viewModel.tvShow!!.voteAverage!!.toFloat(),
                            viewModel.tvShow!!.voteCount!!,
                            viewModel.tvShow!!.overview,
                            viewModel.tvShow!!.posterPath)
                } else {
                    dbService.setFavourite(viewModel.tvShow!!.id)
                }
                fab.setImageResource(R.drawable.ic_stars_black_24dp)
                Snackbar.make(view, "Added to favourite", Snackbar.LENGTH_LONG)
                        .setAction("Action", null).show()
                watched.setImageResource(R.mipmap.checked)
            } else {
                dbService.cancelFavourite(viewModel.tvShow!!.id)
                fab.setImageResource(R.drawable.ic_star_border_black_24dp)
                Snackbar.make(view, "Removed to favourite", Snackbar.LENGTH_LONG)
                        .setAction("Action", null).show()
            }
        }
    }

    private fun getIntentInfo() {
        if (intent != null && viewModel.tvShow == null) {
            val tvShow = TVShow()
            tvShow.id = intent.extras[ID] as Int
            tvShow.name = intent.extras[NAME] as String
            tvShow.voteAverage = intent.extras[VOTE_AVERAGE] as Double
            tvShow.voteCount = intent.extras[VOTE_COUNT] as Int
            tvShow.posterPath = intent.extras[POSTER_PATH] as String
            tvShow.overview = intent.extras[OVERVIEW] as String
            viewModel.tvShow = tvShow

        }
        watched.setImageResource(if (DBMovieService.getInstance().checkIfExists(viewModel.tvShow!!.id)) R.mipmap.checked else R.mipmap.not_checked)
        watched.setOnClickListener {
            if (!DBMovieService.getInstance().checkIfExists(viewModel.tvShow!!.id)) {
                watched.setImageResource(R.mipmap.checked)
                DBMovieService.getInstance().addTVShowToDb(viewModel.tvShow!!.id,
                        viewModel.tvShow!!.title,
                        viewModel.tvShow!!.voteAverage!!.toFloat(),
                        viewModel.tvShow!!.voteCount!!,
                        viewModel.tvShow!!.overview,
                        viewModel.tvShow!!.posterPath
                )
                Snackbar.make(findViewById(R.id.nested_scroll_view), "Added to watched", Snackbar.LENGTH_SHORT).show()
            } else {

                if (DBMovieService.getInstance().checkIfIsFavourite(viewModel.tvShow!!.id)) {
                    Snackbar.make(findViewById(R.id.nested_scroll_view), "It's favourite, u cant do this", Snackbar.LENGTH_SHORT).show()
                } else {
                    val builder = AlertDialog.Builder(this@TVShowDetailsActivity)
                    builder.setMessage(R.string.confirm)
                            .setPositiveButton(R.string.yes) { _, _ ->
                                DBMovieService.getInstance().deleteFromDb(viewModel.tvShow!!.id)
                                watched.setImageResource(R.mipmap.not_checked)
                                Snackbar.make(findViewById(R.id.nested_scroll_view), "Marked as unwatched", Snackbar.LENGTH_SHORT).show()
                            }
                            .setNegativeButton(R.string.no) { _, _ -> }
                    builder.create().show()
                }
            }
        }
        if (DBMovieService.getInstance().checkIfIsFavourite(viewModel.tvShow!!.id)) {
            fab.setImageResource(R.drawable.ic_stars_black_24dp)
        }
    }

    private fun useIntentInfo() {
        toolbar_layout.title = viewModel.tvShow!!.name
        toolbar_layout.setOnClickListener({ zoomImageFromThumb(toolbar_layout, mBitmapDrawable!!) })
        ratingBar.rating = viewModel.tvShow!!.voteAverage!!.toFloat() / 2
        ratingBar.setOnRatingBarChangeListener({ _, rating, fromUser ->  if (fromUser) {
            onInfoUpdatedListener = OnInfoUpdatedListener { code ->
                ratingBar.rating = code
                Snackbar.make(findViewById(R.id.nested_scroll_view), "Your rating saved", Snackbar.LENGTH_LONG).show()
            }
            TVShowsRepository.getInstance().addOnInfoUpdatedListener(onInfoUpdatedListener!!)
            TVShowsRepository.getInstance().rateTVShow(viewModel.tvShow!!.id,rating*2)
        }})
        vote_count.text = "" + Math.round(viewModel.tvShow!!.voteAverage!! * 10).toFloat() / 10 + "/" + viewModel.tvShow!!.voteCount
        overview.text = viewModel.tvShow!!.overview
        release_date.visibility = View.GONE
        share.setOnClickListener{
            val shareLinkContent = ShareLinkContent.Builder()
                    .setQuote(viewModel.tvShow!!.name + "     \r\nPlot: " + viewModel.tvShow!!.overview)
                    .setContentUrl(Uri.parse("https://image.tmdb.org/t/p/w500" + viewModel.tvShow!!.posterPath))
                    .build()
            ShareDialog.show(this@TVShowDetailsActivity, shareLinkContent)}
        Picasso
                .with(this)
                .load("https://image.tmdb.org/t/p/w500" + viewModel.tvShow!!.posterPath)
                .into(object : Target {
                    override fun onBitmapLoaded(bitmap: Bitmap, from: Picasso.LoadedFrom) {
                        mBitmapDrawable = BitmapDrawable(resources, bitmap)
                        mBitmapDrawable!!.gravity = Gravity.NO_GRAVITY
                        toolbar_layout.background = mBitmapDrawable
                    }

                    override fun onBitmapFailed(errorDrawable: Drawable) {
                        Log.d("TAG", "FAILED")
                    }

                    override fun onPrepareLoad(placeHolderDrawable: Drawable) {
                        Log.d("TAG", "Prepare Load")
                    }
                })
    }

    private fun getFullInfo(tvShow: FullTVShow){
        runOnUiThread{
            release_date.visibility = View.VISIBLE
            release_date.text = "" + release_date.text + tvShow.firstAirDate
            if (tvShow.homepage != null && tvShow.homepage != ""){
                links.text = tvShow.homepage
                links.setOnClickListener({
                    val webPage = Uri.parse(tvShow.homepage)
                    val webIntent = Intent(Intent.ACTION_VIEW, webPage)
                    startActivity(webIntent)})
            }
            for (i in tvShow.seasons!!.indices){
                var image = ImageView(this@TVShowDetailsActivity)
                Picasso
                        .with(this@TVShowDetailsActivity)
                        .load("https://image.tmdb.org/t/p/w500" + viewModel.fullTVShow!!.seasons!![i].posterPath)
                        .into(image)
                image.setPadding(0,20,20,0)
                genres.addView(image)
            }
        }
    }

    private fun zoomImageFromThumb(thumbView: View, bitmapDrawable: BitmapDrawable) {
        if (mCurrentAnimator != null) {
            mCurrentAnimator!!.cancel()
        }

        val expandedImageView = findViewById<ImageView>(R.id.expanded_image)
        expandedImageView.setImageDrawable(bitmapDrawable)

        val startBounds = Rect()
        val finalBounds = Rect()
        val globalOffset = Point()

        thumbView.getGlobalVisibleRect(startBounds)
        findViewById<View>(R.id.container).getGlobalVisibleRect(finalBounds, globalOffset)
        startBounds.offset(-globalOffset.x, -globalOffset.y)
        finalBounds.offset(-globalOffset.x, -globalOffset.y)

        val startScale: Float
        if (finalBounds.width().toFloat() / finalBounds.height() > startBounds.width().toFloat() / startBounds.height()) {
            startScale = startBounds.height().toFloat() / finalBounds.height()
            val startWidth = startScale * finalBounds.width()
            val deltaWidth = (startWidth - startBounds.width()) / 2
            startBounds.left -= deltaWidth.toInt()
            startBounds.right += deltaWidth.toInt()
        } else {
            startScale = startBounds.width().toFloat() / finalBounds.width()
            val startHeight = startScale * finalBounds.height()
            val deltaHeight = (startHeight - startBounds.height()) / 2
            startBounds.top -= deltaHeight.toInt()
            startBounds.bottom += deltaHeight.toInt()
        }

        thumbView.alpha = 0f
        expandedImageView.visibility = View.VISIBLE

        expandedImageView.pivotX = 0f
        expandedImageView.pivotY = 0f

        var set = AnimatorSet()
        set
                .play(ObjectAnimator.ofFloat<View>(expandedImageView, View.X, startBounds.left.toFloat(),
                        finalBounds.left.toFloat()))
                .with(ObjectAnimator.ofFloat<View>(expandedImageView, View.Y, startBounds.top.toFloat(),
                        finalBounds.top.toFloat()))
                .with(ObjectAnimator.ofFloat(expandedImageView, View.SCALE_X, startScale, 1f))
                .with(ObjectAnimator.ofFloat(expandedImageView, View.SCALE_Y, startScale, 1f))
        set.duration = mShortAnimationDuration.toLong()
        set.interpolator = DecelerateInterpolator()
        set.addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationEnd(animation: Animator) {
                mCurrentAnimator = null
            }

            override fun onAnimationCancel(animation: Animator) {
                mCurrentAnimator = null
            }
        })
        set.start()
        mCurrentAnimator = set

        expandedImageView.setOnClickListener {
            if (mCurrentAnimator != null) {
                mCurrentAnimator!!.cancel()
            }

            set = AnimatorSet()
            set
                    .play(ObjectAnimator.ofFloat<View>(expandedImageView, View.X, startBounds.left.toFloat()))
                    .with(ObjectAnimator.ofFloat<View>(expandedImageView, View.Y, startBounds.top.toFloat()))
                    .with(ObjectAnimator
                            .ofFloat(expandedImageView, View.SCALE_X, startScale))
                    .with(ObjectAnimator
                            .ofFloat(expandedImageView, View.SCALE_Y, startScale))
            set.duration = mShortAnimationDuration.toLong()
            set.interpolator = DecelerateInterpolator()
            set.addListener(object : AnimatorListenerAdapter() {
                override fun onAnimationEnd(animation: Animator) {
                    thumbView.alpha = 1f
                    expandedImageView.visibility = View.GONE
                    mCurrentAnimator = null
                }

                override fun onAnimationCancel(animation: Animator) {
                    thumbView.alpha = 1f
                    expandedImageView.visibility = View.GONE
                    mCurrentAnimator = null
                }
            })
            set.start()
            mCurrentAnimator = set
        }
    }
}
